### PR TITLE
added OAuth

### PR DESCRIPTION
**Pull request by jygh98**
This PR is regarding #23, we are now able to authenticate users through Google's OAuth API and permit whitelisted users whose added to the database.

### CODE REVIEW

File: ubuntuStation/views.py [hunk 1]
```diff
@@ -1,54 +1,44 @@
...
-    def __init__(self, username):
-        self.id = username
-        self.username = username
-
-    @classmethod
-    def get(cls,id):
-        return cls.user_database.get(id)
+class Config:
+    APP_NAME = "Survey Admin "
+    SECRET_KEY = "development"
```
Comment (carlosparadis) [inline ubuntuStation/views.py:23]: @ryantanaka @jygh98 is this safe to be here? If not, need to get a new secret key and follow the same procedure we did for the other information.

File: ubuntuStation/views.py [hunk 2]
```diff
@@ -289,3 +303,28 @@ def show_questions():
...
+            db_session.rollback()
+            raise
+        finally:
+            db_session.close()
+        return redirect(url_for('show_users'))
+    else:
+        print(form.errors)
+    return render_template('user_form.html', form=form)
+
+@app.route('/showusers', methods=['GET'])
```
Comment (carlosparadis) [inline ubuntuStation/views.py:326]: This should be mentioned in the original issue. I was expecting only the login page to have been changed. Others may not be aware for the same reason.

File: ubuntuStation/templates/base.html [hunk 3]
```diff
@@ -34,6 +34,14 @@
 
 						</ul>
 					</li>
+					<li class="dropdown">
+						<a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">Users <span class="caret"></span></a>
+						<ul class="dropdown-menu" role="menu">
+							<li><a href="{{ url_for('show_users') }}">Manage Users</a></li>
```
Comment (carlosparadis) [inline ubuntuStation/templates/base.html:40]: The "manage users" option added to the menu should be mentioned in the original issue. I was expecting only the login page to have been changed. Others may not be aware for the same reason.

File: ubuntuStation/templates/user_form.html [hunk 4]
```diff
@@ -0,0 +1,24 @@
+{% extends "base.html" %}
+{% block content %}
+<div class="well">
+{% from "_formhelpers.html" import render_field %}
+{{ form.csrf_token }}
+<script src="{{ url_for('static', filename='jquery.sortable.js') }}"></script>
+<h1>Add User</h1>
```
Comment (carlosparadis) [inline ubuntuStation/templates/user_form.html:7]: Adding new users page should be mentioned in the original issue. I was expecting only the login page to have been changed. Others may not be aware for the same reason.

The name of the file also does not portray its purpose. `create_user.html` as the form action name suggests would be better and avoid conflicts in the long run with other file names should we decide to create other forms associated to users.

File: ubuntuStation/views.py [hunk 5]
```diff
@@ -60,38 +50,62 @@ def home():
...
+            print(token)
+            user.tokens = json.dumps(token)
+            db.session.add(user)
+            db.session.commit()
+            login_user(user)
+            return redirect(url_for('home'))
+        return 'Could not fetch your information.'
 
-@app.route('/logoout')
+@app.route('/logout')
```
Comment (carlosparadis) [inline ubuntuStation/views.py:104]: Nice catch.

File: ubuntuStation/models.py [hunk 6]
```diff
@@ -1,39 +1,13 @@
...
-    def verify_password(self, password):
-        return pwd_context.verify(password, self.password_hash)
-
-    def __init__(self, username, password=None):
-        self.id = username
-        self.username = username
-        self.password = password
+from sqlalchemy import PrimaryKeyConstraint
+
+class User(db.Model, UserMixin):
```
Comment (carlosparadis) [inline ubuntuStation/models.py:5]: What tables and columns were modified? Just User?

File: ubuntuStation/templates/login.html [hunk 7]
```diff
@@ -1,22 +1,10 @@
...
-	<div class="form-group">
-		{{ render_field(form.username) }}
-	</div>
-	<div class="form-group">
-		{{ render_field(form.password) }}
-	</div>
-	<button type=submit class="btn btn-primary" value=Submit>Submit</button>
-	</fieldset>
-</form>
+<a href="{{auth_url}}" class="btn";"><img src="static/google_login.png" width="335.2" height="68.8"></a>
```
Comment (carlosparadis) [inline ubuntuStation/templates/login.html:7]: This file change and associated files that enabled `auth_url` are actually the ones truly associated with the title of the PR and Issue (i.e. adds oAuth). Showing users or adding them, albeit related are not required. As such, they need to be explicitly mentioned on issue+pr+commit.

File: run.py [hunk 8]
```diff
@@ -1,6 +1,6 @@
 #!/usr/bin/python3
 from ubuntuStation import app, db
-#from flask import Flask
+import os
 
+os.environ['OAUTHLIB_INSECURE_TRANSPORT'] = '1'
 app.run(debug=True, host='0.0.0.0', port=8989)
```
Comment (carlosparadis) [inline run.py:6]: Is this port here safe (i.e. it is not matching with the apache one, right?)? Also, what is this `insecure transport` equals true about? Looks fishy!

**Review by carlosparadis — CHANGES_REQUESTED** [review]
@jygh98 I've listed inline the requested changes below. **Do not erase the current commit on your fork, or edit it. For now, just create a new commit addressing the issues. This will make it much easier for me to see what you did to fix the pointed problems.**

Only after we agree on the following changes, then we can discuss merging both of them into one. 

Let me know if this doesn't make sense. Worst case scenario you may end up deleting all the review notes I made permanently if you are not on the same page.

**Comment by ryantanaka** [comment]
@ryantanaka

### add oauth
- views.py: remove lines 21-23 (not needed in our case, only for other oauth related features)
- views.py: line 104, uncovered this small mixup here just gonna include it as a separate commit
- models.py: this is to replace the users table (we will need to update this on the database)
- login.html: part of this branch 
- run.py: part of this branch (yes this appears to secure, we've checked some stackoverflow references and will include the link in the PR for reference)

### add  users through survey_admin instead of manually changing the database 
-  views.py: line 323 - 326 can be part of this branch it is a separate feature
- base.html: this relates only to adding oauth users, not to adding oauth
- user_form.html: need to change this to something like create_oauth_user.html

**Comment by carlosparadis** [comment]
The portion associated to Google Login via OAuth has been merged on PR #44

File: ubuntuStation/views.py [hunk 9]
```diff
@@ -60,38 +50,62 @@ def home():
...
+        resp = google.get(Auth.USER_INFO)
+        if resp.status_code == 200:
+            user_data = resp.json()
+            email = user_data['email']
+            user = User.query.filter_by(email=email).first()
+            if user is None:
+                #user = User()
+                #user.email = email
+                return redirect(url_for('login'))
+            user.name = user_data['name']
```
Comment (ryantanaka) [inline ubuntuStation/views.py:95]: @carlosparadis , we don't use user.name anywhere else. I will be removing this. For some reason, `name` was a valid key in the json response that we get, however it no longer is and so we get an error. Fixing it here in 
https://github.com/erdl/survey_admin/issues/64

**Comment by carlosparadis** [comment]
@ryantanaka send on a different PR/issue. Im slightly confused what is this about.

**Comment by ryantanaka** [comment]
yeah i have an issue open already, but im still looking into why this happened.
i wanted to put this comment here so we know when it was added

**Pull request closed without merging**